src/cf.data.pre

```diff
@@ -181,6 +181,13 @@ DOC_START
 	This option is not yet supported by Squid-3.
 DOC_END
 
+# Options removed in 6.x
+NAME: request_entities
+TYPE: obsolete
+DOC_START
+	Remove this line. HTTP/1.1 defines per-message control over GET/HEAD message body using headers.
```

Comment:
This documentation is meant for admins that use an explicit `request_entities` setting today. The current text does not really help the admin to understand what has changed, what the effects of that change are, and how they can tune what they want to tune. Instead of just repeating RFC rules (that have not changed), we should explain why the directive was removed, why the default behavior was changed, and (hint at) how the old behavior can be restored/configured.

Reply:
Over many years the directive has had less and less effect on traffic. With the correct use of Content-Length in HTTP/1.1 it does nothing.

Reply:
> Over many years the directive has had less and less effect on traffic.

Do you mean that the default setting of this directive rejected fewer and fewer legitimate messages? How do we know that?

> With the correct use of Content-Length in HTTP/1.1 it does nothing.

What do you mean? AFAICT, the directive has request rejection effect regardless of whether Content-Length is used correctly (whatever "correct usage" means).

Reply:
> > Over many years the directive has had less and less effect on traffic.
> 
> Do you mean that the default setting of this directive rejected fewer and fewer legitimate messages? How do we know that?

I am estimating from the repeating cycles of people coming to the IETF with specifications to upgrade or replace GET with something that has a body.

Plus the traffic is migrating to HTTP/2 and later which use implicit chunking in the form of DATA frames instead of Content-Length. That seems to be translating into an increase of chunked encoding when they fallback to 1.1. This whole section of code is not run when chunked is used.


> 
> > With the correct use of Content-Length in HTTP/1.1 it does nothing.
> 
> What do you mean? AFAICT, the directive has request rejection effect regardless of whether Content-Length is used correctly (whatever "correct usage" means).

As currently coded in master you are right. However that is counter to a SHOULD in the 1.1 spec - Content-Length with valid value means a body does exist. Once we fix Squid to obey that SHOULD there is no need for admin to turn this on.

Reply:
I am failing to see what you want changed. You asked for "the why" to be added to the line. A brief "why" is already there- that second sentence.

Do we really need to prefix every obsoletion which happens due to RFC compliance updates with "Squid is HTTP/1.1 software and obeys that spec now" ? A longer description placed in release notes often says that, and a more verbose expansion of the "why".

Reply:
> I am failing to see what you want changed.

I want to change the documentation for the removed request_entities directive. The current text does not really help the admin to discover the effects of that change on a Squid instance that used the now-removed option, and to tune Squid configuration to restore or approximate the old behavior (if such tuning is still possible at all). The final text should do that (directly or, if really necessary, by reference to the detailed documentation elsewhere).

> A brief "why" is already there- that second sentence.

I cannot interpret the "HTTP/1.1 defines per-message control over GET/HEAD message body using headers" sentence as a reasonable answer to the "Why the request_entities directive I was using was removed?" question -- the cited specs have not changed, and the directive was actually documented to work around uncertainties in those specs.

And "why?" is not the only question we should answer. In fact, in _this_ context, that "why" question is secondary! The primary questions here are:

* what will change in Squid behavior after the upgrade (assuming the non-default setting was used by the admin prior to the upgrade) and
* how can the admin restore the old behavior (if they can do that at all). 

The "why" questions are primary for the PR description...

This is really basic "document the effect of the change on admins and how to undo it" stuff. We should not be arguing about this at all!

> Do we really need to prefix every obsoletion which happens due to RFC compliance updates with "Squid is HTTP/1.1 software and obeys that spec now"?

That text would be slightly better than the current PR text, IMO, but, in most cases, including this one, we should be more specific. The basic questions are the same for most removals. The answers should be specific to each removal. Most reasonable admins would not know what "Squid is now compliant" mean in the context of this option removal. Heck, some HTTP authors are still arguing what compliance itself means in this area! Squid behavioral changes are specific. We should document them (and any undoing workarounds).

> A longer description placed in release notes often says that, and a more verbose expansion of the "why".

Agreed, but it also sometimes does not. This PR does not contain such a longer description; the effects of this PR has changed quite a bit already and remain very far from obvious; the workarounds, if any, are not obvious either; so I think my request to document the current effects/workarounds is very reasonable.

Reply:
TL;DR: updated the PR description and added release-notes text to this PR.

> > I am failing to see what you want changed.
> 
> I want to change the documentation for the removed request_entities directive. The current text does not really help the admin to discover the effects of that change on a Squid instance that used the now-removed option, and to tune Squid configuration to restore or approximate the old behavior (if such tuning is still possible at all). The final text should do that (directly or, if really necessary, by reference to the detailed documentation elsewhere).
> 

FTR: I believe the spec uncertainty was addressed by the RFC723x editorial work which added the text in RFC 7230 section 3.3 referenced by canUseContentLength() as reason for allowing entity in HTTP/1.1 GET messages.

> > A brief "why" is already there- that second sentence.
> 
> I cannot interpret the "HTTP/1.1 defines per-message control over GET/HEAD message body using headers" sentence as a reasonable answer to the "Why the request_entities directive I was using was removed?" question -- the cited specs have not changed, and the directive was actually documented to work around uncertainties in those specs.

Er, "Because using this directive means you want the behaviour which happens by default now".

> 
> And "why?" is not the only question we should answer. In fact, in _this_ context, that "why" question is secondary! The primary questions here are:
> 
>     * what will change in Squid behavior after the upgrade (assuming the non-default setting was used by the admin prior to the upgrade) and
> 
>     * how can the admin restore the old behavior (if they can do that at all).
> 

The answer to these is many sentences, even paragraphs of descriptive text in the commit description. Not something for a log entry which may be printed whenever Squid reconfigures. A one-liner is enough bloat for the log.

> 
> The "why" questions are primary for the PR description...

Yet this is a request requiring it be placed in the cache.log output. More PR description text I am okay with.

> 
> This is really basic "document the effect of the change on admins and how to undo it" stuff. We should not be arguing about this at all!

So we are to follow GNU project code submission requirements now? That "and how to undo it" is straight from their playbook.

> 
> > Do we really need to prefix every obsoletion which happens due to RFC compliance updates with "Squid is HTTP/1.1 software and obeys that spec now"?
> 
> That text would be slightly better than the current PR text, IMO, but, in most cases, including this one, we should be more specific. The basic questions are the same for most removals. The answers should be specific to each removal. Most reasonable admins would not know what "Squid is now compliant" mean in the context of this option removal. Heck, some HTTP authors are still arguing what compliance itself means in this area! Squid behavioral changes are specific. We should document them (and any undoing workarounds).
> 
> > A longer description placed in release notes often says that, and a more verbose expansion of the "why".
> 
> Agreed, but it also sometimes does not. This PR does not contain such a longer description; the effects of this PR has changed quite a bit already and remain very far from obvious; the workarounds, if any, are not obvious either; so I think my request to document the current effects/workarounds is very reasonable.

That would be cause for a request to update those relevant texts. eg make this PR update the Squid-6 release notes with a text we are both happy with. Not a request to dump lots of info into a log "one-liner".

Reply:
I am resolving this change request. The updated description still needs work, but it is better and further polishing is probably best addressed in a new change request.

> Er, "Because using this directive means you want the behaviour which happens by default now".

You are supposed to tell that to the admin, and the earlier text was not doing that at all. I am saying this here in hope to avoid similar surprising discussions in the future, not to restart this discussion.

> So we are to follow GNU project code submission requirements now? That "and how to undo it" is straight from their playbook.

Strange question. We should require what is right/good/reasonable/etc. I am not familiar with the "GNU project code submission requirements", but if some of our requirements match some of theirs, I see absolutely no problem with that per se. FWIW, the stuff I am requesting is a reasonable minimum IMO.

> Yet this is a request requiring it be placed in the cache.log output.

It is not; you were just misinterpreting it that way. This change request is about providing essential information to the admin. There are many ways to do that.

> More PR description text I am okay with.

A PR description (i.e. commit message) is not something admins can easily find. Anything we want admins to know should be in squid.conf.documented, logs, and/or release notes. Commit messages are for developers.

> That would be cause for a request to update those relevant texts. eg make this PR update the Squid-6 release notes with a text we are both happy with. Not a request to dump lots of info into a log "one-liner".

There is no requirement for the corresponding log message to be limited to one line. It may be possible to limit this particular documentation to one line while addressing my concerns, but if doing so well requires five lines, we should use five lines.